pkg/manager/member/tiflash_scaler_test.go

```diff
@@ -0,0 +1,1739 @@
+
+		if test.hasPVC {
+			pvc1 := newScaleInPVCForStatefulSet(oldSet, v1alpha1.TiFlashMemberType, tc.Name)
+			pvc2 := pvc1.DeepCopy()
+			pvc1.Name = pvc1.Name + "-1"
+			pvc1.UID = pvc1.UID + "-1"
+			pvc2.Name = pvc2.Name + "-2"
+			pvc2.UID = pvc2.UID + "-2"
+			pvcIndexer.Add(pvc1)
+			pvcIndexer.Add(pvc2)
+			pod.Spec.Volumes = append(pod.Spec.Volumes,
+				corev1.Volume{
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc1.Name,
+						},
+					},
+				}, corev1.Volume{
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc2.Name,
+						},
+					},
+				})
+		}
+
+		pod.Labels = map[string]string{}
+		if test.storeIDSynced {
+			pod.Labels[label.StoreIDLabelKey] = "1"
+		}
+		podIndexer.Add(pod)
+
+		pdClient := controller.NewFakePDClient(pdControl, tc)
+
+		pdClient.AddReaction(pdapi.GetConfigActionType, func(action *pdapi.Action) (interface{}, error) {
+			var replicas uint64 = 3
+			return &pdapi.PDConfigFromAPI{
+				Replication: &pdapi.PDReplicationConfig{
+					MaxReplicas: &replicas,
+				},
+			}, nil
+		})
+
+		if test.getStoresFn == nil {
+			test.getStoresFn = func(action *pdapi.Action) (interface{}, error) {
+				store := &pdapi.StoreInfo{
+					Store: &pdapi.MetaStore{
+						StateName: v1alpha1.TiKVStateUp,
+						Store: &metapb.Store{
+							Address: fmt.Sprintf("%s-tiflash-0", "basic"),
+						},
+					},
+				}
+				return &pdapi.StoresInfo{
+					Count:  5,
+					Stores: []*pdapi.StoreInfo{store, store, store, store, store},
+				}, nil
+			}
+		}
+		pdClient.AddReaction(pdapi.GetStoresActionType, test.getStoresFn)
+
+		if test.delStoreErr {
+			pdClient.AddReaction(pdapi.DeleteStoreActionType, func(action *pdapi.Action) (interface{}, error) {
+				return nil, fmt.Errorf("delete store error")
+			})
+		}
+		if test.pvcUpdateErr {
+			pvcControl.SetUpdatePVCError(errors.NewInternalError(fmt.Errorf("API server failed")), 0)
+		}
+
+		err := scaler.ScaleIn(tc, oldSet, newSet)
+		test.errExpectFn(g, err)
+		if test.changed {
+			g.Expect(int(*newSet.Spec.Replicas)).To(Equal(4))
+		} else {
+			g.Expect(int(*newSet.Spec.Replicas)).To(Equal(5))
+		}
+	}
+
+	tests := []testcase{
+		{
+			name:          "store is up, delete store failed",
+			storeFun:      normalTiFlashStoreFun,
+			delStoreErr:   false,
+			hasPVC:        true,
+			storeIDSynced: true,
+			isPodReady:    true,
+			hasSynced:     true,
+			pvcUpdateErr:  false,
+			errExpectFn:   errExpectNotNil,
+			changed:       false,
+		},
+		{
+			name:          "store state is up, delete store success",
+			storeFun:      normalTiFlashStoreFun,
+			delStoreErr:   false,
+			hasPVC:        true,
+			storeIDSynced: true,
+			isPodReady:    true,
+			hasSynced:     true,
+			pvcUpdateErr:  false,
+			errExpectFn:   errExpectRequeue,
+			changed:       false,
+		},
```

Comment:
some fields for these two cases?

Reply:
Oh yeah, maybe the first one can be deleted.